examples/run_summarization_finetuning.py

```diff
@@ -0,0 +1,625 @@
+            with open(cached_features_file, "rb") as source:
+                examples = pickle.load(source)
+                self.source_examples = examples["source"]
+                self.target_examples = examples["target"]
+                return
+
+        logger.info("Creating features from dataset at %s", data_dir)
+        self.source_examples = []
+        self.target_examples = []
+        datasets = ["cnn", "dailymail"]
+        for dataset in datasets:
+            path_to_stories = os.path.join(data_dir, dataset, "stories")
+            assert os.path.isdir(path_to_stories)
+
+            story_filenames_list = os.listdir(path_to_stories)
+            for story_filename in story_filenames_list:
+                path_to_story = os.path.join(path_to_stories, story_filename)
+                if not os.path.isfile(path_to_story):
+                    continue
+
+                with open(path_to_story, encoding="utf-8") as source:
+                    try:
+                        raw_story = source.read()
+                        story_ids, summary_ids = process_story(raw_story, tokenizer)
+                    except IndexError:  # skip ill-formed stories
+                        continue
+
+                story_seq = _fit_to_block_size(story_ids, block_size)
+                self.source_examples.append(story_seq)
+
+                summary_seq = _fit_to_block_size(summary_ids, block_size)
+                self.target_examples.append(summary_seq)
+
+        logger.info("Saving features into cache file %s", cached_features_file)
+        with open(cached_features_file, "wb") as sink:
+            pickle.dump(
+                {"source": self.source_examples, "target": self.target_examples},
+                sink,
+                protocol=pickle.HIGHEST_PROTOCOL,
+            )
+
+    def __len__(self):
+        return len(self.source_examples)
+
+    def __getitem__(self, items):
+        return (
+            torch.tensor(self.source_examples[items]),
+            torch.tensor(self.target_examples[items]),
+        )
+
+
+def process_story(raw_story, tokenizer):
+    """ Extract the story and summary from a story file.
+
+    Attributes:
+        raw_story (str): content of the story file as an utf-8 encoded string.
+
+    Raises:
+        IndexError: If the stoy is empty or contains no highlights.
+    """
+    file_lines = list(
+        filter(lambda x: len(x) != 0, [line.strip() for line in raw_story.split("\n")])
+    )
+
+    # for some unknown reason some lines miss a period, add it
+    file_lines = [_add_missing_period(line) for line in file_lines]
+
+    # gather article lines
+    story_lines = []
+    lines = deque(file_lines)
+    while True:
+        try:
+            element = lines.popleft()
+            if element.startswith("@highlight"):
+                break
+            story_lines.append(element)
+        except IndexError as ie:  # if "@highlight" absent from file
```

Comment:
If « highlight » is absent we will pop elements from the queue even when there are None left. Do not propagate the exception a level above; returning None is more appropriate.